File: src/main/java/com/joecollins/bindings/Bindable.java
package com.joecollins.bindings;

import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.function.Consumer;
import java.util.function.Function;

public abstract class Bindable<T extends Bindable<T, E>, E extends Enum<E>> {

  private Map<E, List<Consumer<T>>> bindings = new ConcurrentHashMap<>();

  public void addBinding(Consumer<T> binding, E... properties) {
    for (E property : properties) bindings.computeIfAbsent(property, createNewList()).add(binding);
  }

  private Function<E, List<Consumer<T>>> createNewList() {
    return x -> new CopyOnWriteArrayList<>();
  }

  public void removeBinding(Consumer<T> binding, E... properties) {
    for (E property : properties)
      bindings.computeIfAbsent(property, createNewList()).remove(binding);
  }

  @SuppressWarnings("unchecked")
  protected void onPropertyRefreshed(E property) {
    bindings
        .getOrDefault(property, Collections.emptyList())
        .forEach(binding -> binding.accept((T) this));
  }
}

File: src/main/java/com/joecollins/bindings/Bindable.kt
package com.joecollins.bindings

import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.CopyOnWriteArrayList

abstract class Bindable<T : Bindable<T, E>, E : Enum<E>> {

    private val bindings: MutableMap<E, MutableList<(T) -> Unit>> = ConcurrentHashMap()

    internal fun addBinding(binding: (T) -> Unit, vararg properties: E) {
        for (property in properties) bindings.computeIfAbsent(property, createNewList()).add(binding)
    }

    private fun createNewList(): (E) -> MutableList<(T) -> Unit> {
        return { CopyOnWriteArrayList() }
    }

    internal fun removeBinding(binding: (T) -> Unit, vararg properties: E) {
        for (property in properties) bindings.computeIfAbsent(property, createNewList()).remove(binding)
    }

    @Suppress("UNCHECKED_CAST")
    protected fun onPropertyRefreshed(property: E) {
        bindings
            .getOrDefault(property, emptyList())
            .forEach { binding -> binding(this as T) }
    }

    internal fun onPropertyRefreshedInternal(property: E) {
        onPropertyRefreshed(property)
    }
}
